testcases/expand.go

```diff
@@ -128,21 +133,91 @@ func applyQueryTweaks(tc *comparer.TestCase, tweaks []*config.QueryTweak) *compa
 	return &resTC
 }
 
+func parseTime(s string) (time.Time, error) {
+	if t, err := strconv.ParseFloat(s, 64); err == nil {
+		s, ns := math.Modf(t)
+		return time.Unix(int64(s), int64(ns*float64(time.Second))).UTC(), nil
+	}
+	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
+		return t, nil
+	}
+	return time.Time{}, errors.Errorf("cannot parse %q to a valid timestamp", s)
+}
+
+// QueryTimeParameters contain endTime, Range (start), and Resolution (Step). If this is given,
+// Then use it to generate v1.Range.
+// Otherwise, we use a hardcoded various values defined in this function.
+func generateQueryTimeRangeParameters(queryTimeParameters config.QueryTimeParameters) []v1.Range {
+	ranges := make([]v1.Range, 0)
+
+	endTimes := make([]time.Time, 0)
+	if queryTimeParameters.EndTime != "" {
+		endTime, err := parseTime(queryTimeParameters.EndTime)
+		if err == nil {
+			endTimes = append(endTimes, endTime)
+		}
+		panic(fmt.Errorf("Invalid config QueryTimeParameters.EndTime received"))
+	} else {
+		// Add an end timestamp that ensures at least some data is on a consuming segment.
+		endTimes = append(endTimes, time.Now().UTC().Add(-2*time.Minute))
+		// Add an end time that ensures at least some data is an offline segment
+		endTimes = append(endTimes, time.Now().UTC().Add(-12*time.Hour))
```

Comment:
is there a way to add a endtime and range which deterministically also covers query covering both realtime and offline in the same query? We did want that case as well right?

Reply:
Ranges more than 6 hours should cover both realtime and offline (and start time of now -2 minutes), since we close the segment every 6 hrs.

Discussed with Pankaj this afternoon. Instead of running all queries with all possible combinations of start/end/ranges, we will just randomly pick from the combinations generated here. There are 900+ queries, so some queries *should* have it. The seed of the random generator is fixed, so the range chosen for a query should be deterministic across runs.